File: rust/bron_kerbosch/src/bron_kerbosch_pivot.rs
//! Core of Bron-Kerbosch algorithms with pivot

use graph::{UndirectedGraph, Vertex, VertexSetLike};
use pile::Pile;
use reporter::Reporter;

use graph::rand::seq::SliceRandom;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PivotChoice {
    Arbitrary,
    Random,
    MaxDegree,
    MaxDegreeLocal,
    MaxDegreeLocalX,
}

type Clique<'a> = Pile<'a, Vertex>;

pub fn visit<VertexSet, Graph, Rprtr>(
    graph: &Graph,
    reporter: &mut Rprtr,
    initial_pivot_selection: PivotChoice,
    further_pivot_selection: PivotChoice,
    mut candidates: VertexSet,
    mut excluded: VertexSet,
    clique: Clique,
) where
    VertexSet: VertexSetLike,
    Graph: UndirectedGraph<VertexSet = VertexSet>,
    Rprtr: Reporter,
{
    debug_assert!(!candidates.is_empty());
    debug_assert!(candidates.all(|&v| graph.degree(v) > 0));
    debug_assert!(excluded.all(|&v| graph.degree(v) > 0));
    debug_assert!(candidates.is_disjoint(&excluded));

    if candidates.len() == 1 {
        // Same logic as below, but stripped down for this common case
        candidates.for_each(|v| {
            let neighbours = graph.neighbours(v);
            if neighbours.is_disjoint(&excluded) {
                reporter.record(clique.place(v).collect());
            }
        });
        return;
    }

    let mut pivot: Option<Vertex> = None;
    let mut remaining_candidates: Vec<Vertex> = Vec::with_capacity(candidates.len());
    match initial_pivot_selection {
        PivotChoice::MaxDegreeLocal | PivotChoice::MaxDegreeLocalX => {
            // Quickly handle locally unconnected candidates while finding pivot
            let mut seen_local_degree = 0;
            candidates.for_each(|v| {
                let neighbours = graph.neighbours(v);
                let local_degree = neighbours.intersection_size(&candidates);
                if local_degree == 0 {
                    // Same logic as below, but stripped down
                    if neighbours.is_disjoint(&excluded) {
                        reporter.record(clique.place(v).collect());
                    }
                } else {
                    if seen_local_degree < local_degree {
                        seen_local_degree = local_degree;
                        pivot = Some(v);
                    }
                    remaining_candidates.push(v);
                }
            });
            if remaining_candidates.is_empty() {
                return;
            }
            if initial_pivot_selection == PivotChoice::MaxDegreeLocalX {
                excluded.for_each(|v| {
                    let neighbours = graph.neighbours(v);
                    let local_degree = neighbours.intersection_size(&candidates);
                    if seen_local_degree < local_degree {
                        seen_local_degree = local_degree;
                        pivot = Some(v);
                    }
                });
            }
        }
        _ => {
            candidates.for_each(|v| remaining_candidates.push(v));
            pivot = choose(initial_pivot_selection, &remaining_candidates, graph).copied();
        }
    }

    debug_assert!(!remaining_candidates.is_empty());
    let pivot = pivot.unwrap();
    for v in remaining_candidates {
        let neighbours = graph.neighbours(v);
        if neighbours.contains(pivot) {
            continue;
        }
        candidates.remove(v);
        let neighbouring_candidates: VertexSet = neighbours.intersection_collect(&candidates);
        if neighbouring_candidates.is_empty() {
            if neighbours.is_disjoint(&excluded) {
                reporter.record(clique.place(v).collect());
            }
        } else {
            let neighbouring_excluded: VertexSet = neighbours.intersection_collect(&excluded);
            visit(
                graph,
                reporter,
                further_pivot_selection,
                further_pivot_selection,
                neighbouring_candidates,
                neighbouring_excluded,
                clique.place(v),
            );
        }
        excluded.insert(v);
    }
}

fn choose<'a, Graph>(
    pivot_choice: PivotChoice,
    candidates: &'a [Vertex],
    graph: &Graph,
) -> Option<&'a Vertex>
where
    Graph: UndirectedGraph,
{
    match pivot_choice {
        PivotChoice::Arbitrary => candidates.first(),
        PivotChoice::Random => {
            let mut rng = rand::thread_rng();
            candidates.choose(&mut rng)
        }
        PivotChoice::MaxDegree => candidates.iter().max_by_key(|&&v| graph.degree(v) as usize),
        _ => panic!("Implemented separately"),
    }
}

File: go/src/bron_kerbosch/bron_kerbosch_pivot.go
package bron_kerbosch

type PivotSelection int

const (
	MaxDegree      PivotSelection = iota
	MaxDegreeLocal                = iota
)

func visit(graph *UndirectedGraph, reporter Reporter,
	initial_pivot_selection PivotSelection, further_pivot_selection PivotSelection,
	candidates VertexSet, excluded VertexSet, clique []Vertex) {
	if len(candidates) == 1 {
		for v := range candidates {
			neighbours := graph.adjacencies[v]
			if excluded.IsDisjoint(neighbours) {
				reporter.Record(append(clique, v))
			}
		}
		return
	}

	var pivot Vertex
	remaining_candidates := make([]Vertex, 0, len(candidates))
	switch initial_pivot_selection {
	case MaxDegree:
		pivot = pick_max_degree(graph, candidates, excluded)
		for v, _ := range candidates {
			remaining_candidates = append(remaining_candidates, v)
		}
		break
	case MaxDegreeLocal:
		// Quickly handle locally unconnected candidates while finding pivot
		var seen_local_degree = 0
		for v := range candidates {
			neighbours := graph.adjacencies[v]
			local_degree := neighbours.IntersectionLen(candidates)
			if local_degree == 0 {
				// Same logic as below, but stripped down
				if neighbours.IsDisjoint(excluded) {
					reporter.Record(append(clique, v))
				}
			} else {
				if seen_local_degree < local_degree {
					seen_local_degree = local_degree
					pivot = v
				}
				remaining_candidates = append(remaining_candidates, v)
			}
		}
		if len(remaining_candidates) == 0 {
			return
		}
		for v := range excluded {
			neighbours := graph.adjacencies[v]
			local_degree := neighbours.IntersectionLen(candidates)
			if seen_local_degree < local_degree {
				seen_local_degree = local_degree
				pivot = v
			}
		}
		break
	}

	for _, v := range remaining_candidates {
		neighbours := graph.adjacencies[v]
		if neighbours.Contains(pivot) {
			continue
		}
		candidates.Remove(v)
		neighbouring_candidates := neighbours.Intersection(candidates)
		if !neighbouring_candidates.IsEmpty() {
			neighbouring_excluded := neighbours.Intersection(excluded)
			visit(
				graph, reporter,
				further_pivot_selection, further_pivot_selection,
				neighbouring_candidates,
				neighbouring_excluded,
				append(clique, v))
		} else {
			if neighbours.IsDisjoint(excluded) {
				reporter.Record(append(clique, v))
			}
		}
		excluded.Add(v)
	}
}

func pick_max_degree(graph *UndirectedGraph, candidates VertexSet, excluded VertexSet) Vertex {
	max_degree := 0
	var max_vertex Vertex
	for v, _ := range candidates {
		degree := graph.degree(v)
		if max_degree < degree {
			max_degree = degree
			max_vertex = v
		}
	}
	return max_vertex
}
